File: ising_wolff.f90
program ising_wolff
    implicit none
    
    !------------------------------------------------------------------------!
    !                               parameters                               !
    !------------------------------------------------------------------------!
    integer, parameter :: N = 1e5                ! Monte Carlo steps
    integer, parameter :: L = 50,nx = L, ny = L   ! square lattice size
    real(8), parameter :: T = 2.2d0               ! temperature
    real(8), parameter :: p = 1.0d0-exp(-2.0d0/T) ! bonds probability
    
    !------------------------------------------------------------------------!
    !                               variables                                !
    !------------------------------------------------------------------------!
    integer :: it, i, j, Si, n_add, ic
    real(8) :: r
    integer, dimension(nx,nx) :: S ! array of spins
    logical, dimension(nx,ny) :: C ! array of clustered spins
    integer, dimension(2,4) :: s_add

    !------------------------------------------------------------------------!
    !                        Monte Carlo Wolff Method                        !
    !------------------------------------------------------------------------!

    call initial_state(S) ! asign random values to the spins (1 or -1)

    do it = 1,N
        C = .FALSE.
        call random_spin(i,j) ! randomly choose a spin
        C(i,j) = .TRUE.
        Si = S(i,j) ! value of the choosen spin
        call cluster_formation(i,j,n_add,s_add,C)
        do while (n_add > 0)
            do ic = 1,n_add
                Si=S(s_add(1,ic),s_add(2,ic))
                call cluster_formation(s_add(1,ic),s_add(2,ic),n_add,s_add,C)
            end do
        end do
        do i = 1,nx
            do j = 1,ny
                if (C(i,j)) S(i,j) = -S(i,j)
            end do 
        end do 
    end do

    open (unit = 100, file = "final_state.txt", status = "replace")
        do i = 1,L
            do j = 1,L
                write(100,*) S(i,j)
            end do
        end do
    close(100)

    !------------------------------------------------------------------------!
    !                        functions and subroutines                       !
    !------------------------------------------------------------------------!

    contains

    ! random initial state
    subroutine initial_state(S)
        integer, intent(out) :: S(nx,ny)
        open (unit = 100, file = "initial_state.txt", status = "replace")
            do i = 1,L
                do j = 1,L
                    call random_number(r)
                    S(i,j) = 2*int(r+0.5d0)-1
                    write(100,*) S(i,j)
                end do
            end do
        close(100)
    end subroutine initial_state

    ! random initial state
    subroutine random_spin(i,j)
        integer, intent(out) :: i, j
        call random_number(r)
        i = 1+int(r*nx)
        call random_number(r)
        j = 1+int(r*ny)
    end subroutine random_spin

    ! neighbors with periodic boudary conditions for cluster formation
    subroutine cluster_formation(i,j,n_add,s_add,C)
        integer, intent(in) :: i, j
        integer, intent(inout) :: s_add(2,4), n_add
        logical, intent(inout) :: C(nx,ny)
        integer :: ip, im, jp, jm

        ip = mod(i,nx)+1
        im = mod(i-2,nx)+1
        jp = mod(j,ny)+1
        jm = mod(j-2,ny)+1

        n_add = 0
        call random_number(r)
        if (S(ip,j) == Si .AND. r < p .AND. .NOT. C(ip,j)) then
            n_add = n_add+1
            s_add(1,n_add) = ip
            s_add(2,n_add) = j
            C(ip,j) = .TRUE.
        end if
        call random_number(r)
        if (S(im,j) == Si .AND. r < p .AND. .NOT. C(im,j)) then
            n_add = n_add+1
            s_add(1,n_add) = im
            s_add(2,n_add) = j
            C(im,j) = .TRUE.
        end if
        call random_number(r)
        if (S(i,jp) == Si .AND. r < p .AND. .NOT. C(i,jp)) then
            n_add = n_add+1
            s_add(1,n_add) = i
            s_add(2,n_add) = jp
            C(i,jp) = .TRUE.
        end if
        call random_number(r)
        if (S(i,jm) == Si .AND. r < p .AND. .NOT. C(i,jm)) then
            n_add = n_add+1
            s_add(1,n_add) = i
            s_add(2,n_add) = jm
            C(i,jm) = .TRUE.
        end if
    end subroutine cluster_formation

end program ising_wolff

File: C/temp_v_mag/ising_wolff.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>
#include <time.h>

/*---------------------------------------------------------------------------*\
|                                Parameters                                   |
\*---------------------------------------------------------------------------*/ 

#define N 1000                                   // monte Carlo steps
#define L 250                                   // square lattice size
#define nx L
#define ny L
#define Ti 0.01                                   // temperature
#define Tf 4.0
#define nt 100
#define f 1                                     // saving data frequency

/*---------------------------------------------------------------------------*\
|                              Global variables                               |
\*---------------------------------------------------------------------------*/

int S[nx][ny];          // array of spins
bool C[nx][ny];         // array of clustered spins
int s_add[2][nx * ny];  // array to store cluster positions

// functions
void initial_state();
void random_spin(int *i, int *j);
void cluster_formation(int Si, double p, int i, int j, int *n_add);

int main() {
    int i, j, it, Si, n_add, ic, E, M, ip, jp;
    double T, p;
    FILE *state_file, *mag_file, *energy_file;

    srand(time(NULL));

    // initial state
    initial_state();

    // open files
    mag_file = fopen("data.txt", "w");

    for (int ite = 0; ite <nt; ite++) {

        T = Ti+ite*(Tf-Ti)/nt;
        p = 1.0-exp(-2.0/T);

        M = 0;

        for (it = 1; it <= N; it++) {

            // reset cluster array
            for (i = 0; i < nx; i++) {
                for (j = 0; j < ny; j++) {
                    C[i][j] = false;
                }
            }

            // choose a random spin
            random_spin(&i,&j);
            C[i][j] = true;
            Si = S[i][j];
            n_add = 1;
            s_add[0][0] = i;
            s_add[1][0] = j;

            // form cluster
            for (ic = 0; ic < n_add; ic++) {
                cluster_formation(Si, p, s_add[0][ic], s_add[1][ic], &n_add);
            }

            // flip the cluster
            for (i = 0; i < nx; i++) {
                for (j = 0; j < ny; j++) {
                    if (C[i][j]) {
                        S[i][j] = -S[i][j];
                    }
                }
            }
        }
        M = 0;
        for (i = 0; i < nx; i++) {
            for (j = 0; j < ny; j++) {
            M += S[i][j];
            }
        }
        fprintf(mag_file, "%d %f\n", M, T);
    }

    // close files
    fclose(mag_file);
    return 0;
}

/*---------------------------------------------------------------------------*\
|                                  Functions                                  |
\*---------------------------------------------------------------------------*/

// initial state with random spins
void initial_state() {
    int i, j;
    for (i = 0; i < nx; i++) {
        for (j = 0; j < ny; j++) {
            S[i][j] = (rand()%2)*2-1; // Randomly assign -1 or 1
        }
    }
}

// randomly choose a spin
void random_spin(int *i, int *j) {
    *i = rand()%nx;
    *j = rand()%ny;
}

// form cluster with periodic boundary conditions
void cluster_formation(int Si, double p, int i, int j, int *n_add) {
    int ip, im, jp, jm;
    double r;

    ip = (i+1)%nx;
    im = (i-1+nx)%nx;
    jp = (j+1)%ny;
    jm = (j-1+ny)%ny;

    r = (double)rand()/RAND_MAX;
    if (S[ip][j] == Si && r < p && !C[ip][j]) {
        s_add[0][*n_add] = ip;
        s_add[1][*n_add] = j;
        C[ip][j] = true;
        (*n_add)++;
    }

    r = (double)rand()/RAND_MAX;
    if (S[im][j] == Si && r < p && !C[im][j]) {
        s_add[0][*n_add] = im;
        s_add[1][*n_add] = j;
        C[im][j] = true;
        (*n_add)++;
    }

    r = (double)rand()/RAND_MAX;
    if (S[i][jp] == Si && r < p && !C[i][jp]) {
        s_add[0][*n_add] = i;
        s_add[1][*n_add] = jp;
        C[i][jp] = true;
        (*n_add)++;
    }

    r = (double)rand()/RAND_MAX;
    if (S[i][jm] == Si && r < p && !C[i][jm]) {
        s_add[0][*n_add] = i;
        s_add[1][*n_add] = jm;
        C[i][jm] = true;
        (*n_add)++;
    }
}
